client/beefy/src/tests.rs

```diff
@@ -0,0 +1,442 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Tests and test helpers for BEEFY.
+
+use futures::{future, stream::FuturesUnordered, Future, FutureExt, StreamExt};
+use parking_lot::{Mutex, RwLock};
+use std::{pin::Pin, sync::Arc, task::Poll, time::Duration};
+use tokio::runtime::{Handle, Runtime};
+
+use sc_consensus::{BoxJustificationImport, LongestChain};
+use sc_finality_grandpa::{
+	block_import, run_grandpa_voter, Config, GenesisAuthoritySetProvider, GrandpaParams, LinkHalf,
+	SharedVoterState,
+};
+use sc_keystore::LocalKeystore;
+use sc_network::config::{ProtocolConfig, Role};
+use sc_network_test::{
+	Block, BlockImportAdapter, FullPeerConfig, PassThroughVerifier, Peer, PeersClient,
+	PeersFullClient, TestNetFactory,
+};
+
+use beefy_primitives::{
+	crypto::AuthorityId as BeefyId, BeefyApi, ValidatorSet as BeefyValidatorSet,
+	KEY_TYPE as BeefyKeyType,
+};
+use sp_api::{ApiRef, ProvideRuntimeApi};
+use sp_application_crypto::key_types::GRANDPA;
+use sp_blockchain::Result;
+use sp_finality_grandpa::AuthorityList;
+use sp_keyring::Ed25519Keyring;
+use sp_keystore::{SyncCryptoStore, SyncCryptoStorePtr};
+use sp_runtime::{generic::BlockId, traits::Header as HeaderT};
+
+use crate::{keystore::tests::Keyring as BeefyKeyring, notification::*};
+
+const GRANDPA_PROTOCOL_NAME: &'static str = "/grandpa/1";
+const BEEFY_PROTOCOL_NAME: &'static str = "/beefy/1";
+
+type BeefyPeer = Peer<PeerData, GrandpaBlockImport>;
+type GrandpaBlockImport = sc_finality_grandpa::GrandpaBlockImport<
+	substrate_test_runtime_client::Backend,
+	Block,
+	PeersFullClient,
+	LongestChain<substrate_test_runtime_client::Backend, Block>,
+>;
+
+pub struct BeefyLinkHalf {
+	signed_commitment_stream: BeefySignedCommitmentStream<Block>,
+	beefy_best_block_stream: BeefyBestBlockStream<Block>,
+}
+type GrandpaLinkHalf =
+	LinkHalf<Block, PeersFullClient, LongestChain<substrate_test_runtime_client::Backend, Block>>;
+
+#[derive(Default)]
+struct PeerData {
+	grandpa_link_half: Mutex<Option<GrandpaLinkHalf>>,
+	beefy_link_half: Mutex<Option<BeefyLinkHalf>>,
+}
+
+struct BeefyTestNet {
+	peers: Vec<BeefyPeer>,
+	test_config: TestCfg,
+	test_api: Arc<TestApi>,
+}
+
+impl BeefyTestNet {
+	fn new(test_config: TestCfg, n_authority: usize, n_full: usize) -> Self {
+		let mut net = BeefyTestNet {
+			peers: Vec::with_capacity(n_authority + n_full),
+			test_config,
+			test_api: Arc::new(TestApi {}),
+		};
+
+		for _ in 0..n_authority {
+			net.add_authority_peer();
+		}
+
+		for _ in 0..n_full {
+			net.add_full_peer();
+		}
+
+		net
+	}
+
+	fn add_authority_peer(&mut self) {
+		self.add_full_peer_with_config(FullPeerConfig {
+			notifications_protocols: vec![GRANDPA_PROTOCOL_NAME.into()],
```

Comment:
Adding beefy proto name here seems to break grandpa voting...

```suggestion
			notifications_protocols: vec![GRANDPA_PROTOCOL_NAME.into(), BEEFY_PROTOCOL_NAME.into()],
```